### ISSUE TITLE
Proposal: limited enrolling slots

### ISSUE THREAD

**Issue by Beanow** [micromdm/micromdm#320]
To restrict how a device might be enrolled, using a concept of slots.
(#310 / #81)

1. Call the API commands to create a new unique slot.
    Produces an enroll link for this slot `/mdm/enroll/<Slot ID>`.
    A random SCEP challenge is created and included.
    Slot is stored in the DB with the ID, the **served** state.
1. The profile served on that link will contain the same Slot ID on `/mdm/checkin/<Slot ID>`.
    In this mode, requests without a known Slot ID will be rejected.
    If accepted the enrolling proceeds as normal.
    Slot is updated in the DB to have the UDID and the **consumed** state.

An additional API endpoint and mdmctl command will allow itterating all the slots.

This will allow an implementer to have an "invite-only" scheme where your application can decide what preconditions will warrant a slot to be created.

And after the fact this slot can be associated with a device once enrolling has completed. Allowing the application to tie this device into a broader application context.

**Comment by groob** [comment]
That's an interesting idea. We definitely need to address who is allowed to enroll and have discussed several possible ways of doing it, but never this. 

Let's wait and see what input comes from other users/devs.

**Comment by Beanow** [comment]
Although I'm guessing at their implementation details, it's a system that https://simplemdm.com/ applies to associate a particular device with a user.

**Comment by groob** [comment]
@jessepeterson @bruienne @np5 I'd be curious as to thoughts about this proposal.

**Comment by jessepeterson** [comment]
**tl;dr:** yes; we've talked about systems like this and their benefits. as always devil's in the details. 😄 

This is essentially the idea with OTA enrollments and why that stub code exists. Your Slot ID is essentially what we've been calling a token all along. How it initially gets "issued" would be a matter up to the admin/implementor.

With this specific proposal I'm just not sure why the slot is passed around so often in the URL endpoints? On each request to CheckInURL and Command endpoints you are authenticated against the client certificate. If the client certificate is generated from the SCEP request the only real place this "slot" (or token) is needed is in the SCEP challenge field of the enrollment profile (i.e. there's an inherent association between the device cert and the SCEP challenge). No need to pass it around on the endpoints. Of course if you don't have control of the SCEP server you may not be able to get that association (but perhaps we proxy SCEP in this case to capture issued certificates). But if you don't have access to the SCEP server "unique" challenges may not be possible anyway.

Since manual MDM enrollment is less and less useful these days and DEP profile assignment is the way of the future we'll need a token/slot auto-generation mechanism. And under those conditions: what purpose is it then serving? One consideration for these tokens that's been mentioned (e.g. in #81) is re-usable enrollment profiles for mass/automated enrollment (and/or  Configurator workflows). I'm not sure how tokens/slots will work in that workflow (other than, say, OTA profiles).

There's further things around tokens/slots too such as timeouts, and clean-up and other misc. details there. One-time use cases and token/timeouts are also something we're considering to protecte `InstallApplication` download URLs.

**Comment by groob** [comment]
A proposal I prefer is to allow enrollment by default, but but the device into a insecure group. 

The reason is that denying enrollments is not a good UX by any means. And you get to miss the DEP bootstrap if you deny enrollment. 

But allowing enrollments and validating the user in a login session right now is more viable. 

I'm experimenting with killing the setup assistant before the login session comes up. I suspect that a workflow that combines NoMAD and DEP could prove to be quite powerful here. 

As an aside, micromdm/scep does support both dynamic challenges like an NDES server and autosign similar to how puppet cert autosigning works. Both of these features will make their way to micromdm scep at some point.

**Comment by bruienne** [comment]
I like this idea. Like @groob said, combining the two with a general "unprivileged" enrollment ability that can be flipped to a trusted enrollment based on user or other attestation methods one could trigger a first boot enrollment tool like NoMAD or some extension of tools like SplashBuddy to offer a full onboarding and setup experience. This also allows us to care less about Apple's limited initial auth methods (Basic or Digest) by implementing a modern SSO+2FA driven user onboarding experience that runs afterwards.

**Comment by erikng** [comment]
@clburlison has some code for an unprivileged dep enrollment. You should ask him to show you @groob

**Comment by groob** [comment]
I'd love to hear what Clay has to say as well :)

**Comment by Beanow** [comment]
@jessepeterson 
> I'm just not sure why the slot is passed around so often in the URL endpoints?

For `/mdm/enroll` it's simply for a early rejection and templating the profile.

For `/mdm/checkin` this way we can limit the scope and security properties of the slot IDs.

By it's nature this slot ID needs to leave the well defined enrolling context and will be sent by whatever means, email, sms, chat, QRs,... to the device. So the implications of these transportation means might compromise the SCEP flow. I would suggest the SCEP challenge is not the same as slot ID.

Beyond that, in order to have this association property (Slot ID => UDID), we would need specific insights into how the SCEP implementation works to use the challenge for this. When this is done by an external SCEP or delegated in a high availability context this becomes needlessly complicated versus simply including it in the checkin URL. The entire slots concept would then have no impact on SCEP.

> Since manual MDM enrollment is less and less useful these days and DEP profile assignment is the way of the future

In our use-case the devices are not newly purchased, not corporate owned, and the setup is not done by employees. The initiator of the enrolling process is an external customer with their devices. This is why we need both a manual and associative flow. This way an already authenticated customer will own the Slot ID, to be later associated with a device. DEPs constructs do not apply to this afaik.

> There's further things around tokens/slots too such as timeouts, and clean-up and other misc. details there.

Indeed that implies some work. However it's exactly because those details are fairly intricate with the enrolling flow itself that I propose this is implemented by MicroMDM and not some proxy / application on top.

**Comment by Beanow** [comment]
> A proposal I prefer is to allow enrollment by default, but but the device into a insecure group.

This is something you can implement today if you're willing to use the API. Simply treat the entire pool of devices in MicroMDM as the insecure group and define the secured ones in your application.

Speaking of defaults however, I failed to mention this but.
Using slots is backwards incompatible as it breaks existing manual enrolling flows.
This means the slot based flow must be optional and off by default. I would implement this as a flag like `-require-slots true` for the serve command.

So as your proposal is already possible and the new idea is optional, isn't that the best of both? :smile:

**Comment by Beanow** [comment]
@jessepeterson I wasn't aware of the OTA details.
You mean this right? https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/iPhoneOTAConfiguration/OTASecurity/OTASecurity.html#//apple_ref/doc/uid/TP40009505-CH3-SW1

Can you link me to any discussions / design docs about that for MicroMDM? I'd like to see if that fits the purpose.

**Comment by np5** [comment]
We have implemented the MDM OTA enrollment in Zentral – The MDM in Zentral is an experiment to get the kernel extensions to work in 10.13, and to see how the enrollments can be secured via SCEP and some challenges. We hope to be able to integrate micromdm for that, but cannot at the moment, because of the limited ways to verify and monitor the enrollment process.

Here is how we do it.

We have the concept of [enrollment secrets](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/inventory/models.py#L810), which can be limited in time, limited to specific devices, revoked, and which generate events when they are used and verified. They are linked to a business unit too, to be able to classify the devices automatically.

PHASE 1

First, we use one of this secret to start an enrollment "campaign" and generate a profile service payload. [The Challenge in the profile service payload](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/payloads.py#L90) is the secret of the enrollment secret.

[PHASE 2](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/views/ota.py#L52)

A device contacts the enrollment endpoint, with a payload signed by an (expired?, unknown?) apple certificate. Zentral can verify that the Challenge is still valid (not revoked, limited to some serial numbers or not, ...), and then start an ["enrollment session"](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/models.py#L94), with a "Phase 2" status.

The enrollment session is there to associate a device with an enrollment campaign, and follow the progress of the enrollment. It contains a new secret, that can only be used twice by a machine **with the serial number and UDID provided in the first OTA enrollment request**.

The secret is used to generate a SCEP payload. It is placed in the common name of the subject. The serial number is added to the subject, as well as the business unit (Organization).

The device builds a CSR with the information contained in the SCEP payload and contacts the SCEP server. With the newly added external verification in micromdm/scep, we fork a process that posts the CSR to Zentral. Zentral can [verify the content, identify the device, update the enrollment session status ("Phase 2 SCEP verified")](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/views/setup.py#L184).

[PHASE 3](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/views/ota.py#L77)

The device contacts the OTA enrollment endpoint again, this time with a payload signed by the newly procured certificate. We verify the signature, the enrollment session (CN of the certificate). The status of the enrollment session is updated ("Phase 3"). A response is built, containing a root certificate payload, a new SCEP payload and the MDM profile. 

This is the end of the OTA phase of the enrollment. Three more requests are necessary for a full MDM enrollment.

The device then requests the definitive certificate (last SCEP payload). Again, the CSR is checked by Zentral, moving the enrollment session to the next status ("Phase 3 SCEP verified"), and closing the secret that has now been used twice.

The device then contacts the MDM/Checkin endpoint, with an Authenticate request, using the new certificate. Nginx can verify the certificate and pass the request to Zentral, with the client certificate subject in an extra header. Zentral can verify the enrollment session (CN of the client cert), and [update its status](https://github.com/zentralopensource/zentral/blob/2dae1267af12d518ec5de8a07802160bc9b666bd/zentral/contrib/mdm/views/mdm.py#L101) to "Authenticated".

Finally, the device contacts the MDM/Checkin endpoint, with a Token Update request. Again, Nginx and Zentral can verify the request. The enrollment session moves to the last status "Completed".

We still haven't tried to use these concepts for a DEP enrollment, but some parts could definitely be reused.

I hope this experiment / proof of concept can help.

**Comment by groob** [comment]
@np5 that all sounds good, but 
1) how do you validate the user during DEP without blocking their current enrollment?
2) what is the process like for verifying that the whitelist you have for who can enroll? is there a lot of operational overhead?

**Comment by jessepeterson** [comment]
@Beanow

Yes, that's the OTA enrollment workflow.

> Can you link me to any discussions / design docs about [OTA enrollment] for MicroMDM? I'd like to see if that fits the purpose.

Our implementation of OTA enrollment right now doesn't solve any particular issue (other than guarding the enrollment profile from being only accessible from Apple devices—much as DEP enrollment does). It was implemented as more of an exploration of what OTA is and how it works. In the end the same ol' enrollment profile is delivered.

However, because of the nature of OTA's two-step "enrollment" guarding the enrollment profile it is possible to embed information in the enrollment profile (like SCEP challenges, enrollment URL details, etc.) that can be somewhat protected. It is a bit of security-through-obscurity, but it at least serves that purpose.

There hasn't been much talk about specifically how the OTA enrollment fits in with MicroMDM. Largely because there's a backlog of work :) like user support and authentication backends and such that could facilitate something like, say, SAML authentication before OTA profile delivery, token/slot discussions like the above, etc. But that's the general concept in which it was explored for. Hope that's helpful!

**Comment by jessepeterson** [comment]
@np5 Thanks for that explanation! You kindly linked to the code, too, but getting the overview was nice. Thanks! @Beanow workflows like @np5 described are exactly the type of thing the OTA enrollment provides (or at least assists in).

**Comment by jessepeterson** [comment]
Finally, one last thing: in certain environments I absolutely think its fine to block an enrollment if your security policy says so. Regardless of UX. I like the 'unprivileged' enrollment idea as a default, but some environments may just require blocking of enrollment.

**Comment by Beanow** [comment]
@jessepeterson 

> For what it's worth Apple's MDM agreements specifically forbid doing this.

https://simplemdm.com/ doesn't own the devices their customers enroll, how do they not violate those terms? Thanks for the heads up though, we'll have a look at what it says.

@np5's Zentral example looks really sensible to me.

The core being the pre-generated secret, which gets checked as the challenge at various stages.
It doesn't seem to require OTA, if you don't need tailored profiles based on the validation results and don't require the device info (UDID, Serial, IMEI, etc.) for your validation logic.

So I think this could be implemented separately from what the final utility of OTA will be in MicroMDM and is applicable to non-OTA enrollments too.

We could also delegate the majority of the state tracking to an external service, so long as MicroMDM does a small amount of handling the challenges to include them in profiles.

@jessepeterson indeed looking further into it, going through SCEP looks like the better approach.

I'll post a revision to reflect the feedback.

**Comment by Beanow** [comment]
Since @np5's usecase is a bit more involved it perhaps doesn't fit the slot term, but sticking to it for lack of a better term.
Token is better than ID though as it could be some other unique secret or even JWTs for all I care. It's an opaque string to MicroMDM/SCEP though.
Here's what I think a tracker abstraction should implement.

## External slot tracker

**Modes**

- manual
- ota

**States**

1. created
1. served
1. attrs_validated
1. scep_signed
1. phase3_served
1. enrolled
1. expired

**ValidateToken(mode, targetState, slotToken) (bool, err)**

Tells us whether a matching slot token exists for a given mode and _target_ state.

Your validation logic may ignore or combine modes and/or states as needed.
Your validation logic may also choose whether to silently reject with a false value, or provide feedback through an error.

**ValidateDeviceAttrs(mode, deviceAttrs, slotToken) (bool, err)**

Very similar to ValidateChallenge, except here the attrs_validated state is implied.

**RequestStateChange(targetState, slotToken) (bool, err)**

This is a way to complete steps of the process.
Note it's a single direction update process, e.g. a `scep_signed` state can't go back to `created`.

For Manual enrolling:

1. External logic produces a new slot. state: `created`
1. The profile is served. state: `served`
1. SCEP validates the slot token and signs. state: `scep_signed`
1. Device checks in with Authenticate message. state: `enrolled`

For OTA enrolling:

1. External logic produces a new slot. state: `created`
1. The Phase 1 payload is served. state: `served`
1. The Device provides it's attrs and Phase 2 profile is served. state: `attrs_validated`
1. SCEP validates the slot token and signs. state: `scep_signed`
1. Device requests Phase 3 and is served the MDM profile. state: `phase3_served`
1. Device checks in with Authenticate message. state: `enrolled`

## Changes required

**MicroMDM**

The `/mdm/enroll` and `/ota/enroll` endpoints should accept slot tokens.
The tokens should be validated through the external tracker before completing the endpoint's actions.
Profiles need to be modified to include the token (whether custom or a generated one).
State changes need to be requested when done.

The `/mdm/checkin` endpoint also needs to do a best-effort validation and request state changes.
Note that unenrolling also requests the expired state so a new slot is required for re-enrolling.

**SCEP**

@np5 is working around the challenge not being proxied by including it in the certificate and proxying the CSR.
Either SCEP should proxy the challenge as well or implement the same external slot tracker features.

The token should be taken from the challenge and be validated before signing.
A state change need to be requested when done.

**Comment by jessepeterson** [comment]
@Beanow

> https://simplemdm.com/ doesn't own the devices their customers enroll, how do they not violate those terms? Thanks for the heads up though, we'll have a look at what it says.

Oh, sorry, maybe I misunderstood what you were saying. SimpleMDM doesn't own the devices, but the people enrolling devices can't enroll devices that they do not own. Does that make sense? I can dig through the various EULAs to find the specific text if you'd like (just have to give me some time).

**Comment by Beanow** [comment]
@jessepeterson oh I see, a customer can enroll their phones, just not their neighbors phones. That should be no problem.
